### PR TITLE
add Helix cloud interface and implementation skeleton methods

### PR DESCRIPTION
**Pull request by zhangmeng916**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:
issue:[599](https://github.com/apache/helix/issues/599)

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
We plan to support participant auto registration to the cluster for scalability and robustness reason. In this PR, there are interfaces for fetching and parsing cloud information, and some skeleton code for the implementation.

### Tests

- [x] The following is the result of the "mvn test" command on the appropriate module:

[ERROR] Tests run: 882, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 3,152.768 s <<< FAILURE! - in TestSuite
[ERROR] testTaskPerformanceMetrics(org.apache.helix.monitoring.mbeans.TestTaskPerformanceMetrics)  Time elapsed: 2.16 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>
	at org.apache.helix.monitoring.mbeans.TestTaskPerformanceMetrics.testTaskPerformanceMetrics(TestTaskPerformanceMetrics.java:118)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestTaskPerformanceMetrics.testTaskPerformanceMetrics:118 expected:<true> but was:<false>
[INFO] 
[ERROR] Tests run: 882, Failures: 1, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Apache Helix 0.9.2-SNAPSHOT:
[INFO] 
[INFO] Apache Helix ....................................... SUCCESS [  0.388 s]
[INFO] Apache Helix :: Core ............................... FAILURE [52:36 min]
[INFO] Apache Helix :: Admin Webapp ....................... SKIPPED
[INFO] Apache Helix :: Restful Interface .................. SKIPPED
[INFO] Apache Helix :: HelixAgent ......................... SKIPPED
[INFO] Apache Helix :: Recipes ............................ SKIPPED
[INFO] Apache Helix :: Recipes :: Rabbitmq Consumer Group . SKIPPED
[INFO] Apache Helix :: Recipes :: Rsync Replicated File Store SKIPPED
[INFO] Apache Helix :: Recipes :: distributed lock manager  SKIPPED
[INFO] Apache Helix :: Recipes :: distributed task execution SKIPPED
[INFO] Apache Helix :: Recipes :: service discovery ....... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  52:37 min
[INFO] Finished at: 2019-11-13T10:12:30-08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M3:test (default-test) on project helix-core: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/mnzhang/helix/helix-core/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :helix-core

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Code Quality

- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/cloud/AzureComputeInfo.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package org.apache.helix.cloud;
```
Comment (lei-xia) [inline helix-core/src/main/java/org/apache/helix/cloud/AzureComputeInfo.java]: I suggest to add a sub-module call azure for all Azure implementations.

File: helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationFetcher.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * An interface that provides cloud information fetching
+ * @param <T> The type of record to use
+ */
+public interface CloudInformationFetcher<T> {
```
Comment (lei-xia) [inline helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationFetcher.java]: And be more accurate, this is  getVMInformation, right?

File: helix-core/src/main/java/org/apache/helix/cloud/AzureInformationFetcher.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.helix.common.cloud.CloudInformationFetcher;
+
+public class AzureInformationFetcher implements CloudInformationFetcher {
```
Comment (lei-xia) [inline helix-core/src/main/java/org/apache/helix/cloud/AzureInformationFetcher.java]: VMInformationFetcher? or more generic, InstanceInformationFetcher?

File: helix-core/src/main/java/org/apache/helix/cloud/AzureInformationFetcher.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  private final String _azureUrl;
+
+  public AzureInformationFetcher(String azureUrl) {
```
Comment (lei-xia) [inline helix-core/src/main/java/org/apache/helix/cloud/AzureInformationFetcher.java]: be more specific? azureUrl is confusing? what does that mean?

File: helix-core/src/main/java/org/apache/helix/cloud/AzureInformationFetcher.java (hunk 3 above)
Comment (lei-xia) [inline helix-core/src/main/java/org/apache/helix/cloud/AzureInformationFetcher.java]: Please add JAVADOC to describe the  function/features of  these classes and interfaces.  ALso javaDoc for public  methods.

File: helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationFetcher.java (hunk 2 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationFetcher.java]: In the interface, we don't want to be very specific, right? For Azure, the response actually has more than VM information. It also includes VMSS info, location info, network info etc. We can parse the response and get the fields we're interested in.

File: helix-core/src/main/java/org/apache/helix/cloud/AzureInformationFetcher.java (hunk 3 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/cloud/AzureInformationFetcher.java]: Please see above comment. I'm thinking whether we should keep it more general than VM/Instance, so later if we want to get other field's information, we can also use this function.

File: helix-core/src/main/java/org/apache/helix/cloud/AzureInformationFetcher.java (hunk 4 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/cloud/AzureInformationFetcher.java]: It's a static url for requesting instance information.
http://169.254.169.254/metadata/instance
I'll think about a better naming.

File: helix-core/src/main/java/org/apache/helix/cloud/AzureInformationFetcher.java (hunk 3 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/cloud/AzureInformationFetcher.java]: Rethink about it, and this API is actually called instance metadata service. Maybe we should have something similar in the name.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataFetcher.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+public class AzureInstanceMetadataFetcher implements CloudInformationFetcher {
+  private final String _azureUrl;
+
+  public AzureInstanceMetadataFetcher(String azureInstanceMetadataEndpoint) {
+    _azureUrl = azureInstanceMetadataEndpoint;
+  }
+
+  /**
+   * Get the instance metadata information from Azure using the azure instance metadata endpoint
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataFetcher.java]: Let's make it clear that it's the VM instance metadata info?

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParseException;
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java]: Question: Why Gson over Jackson? Helix already uses Jackson.

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java [hunk 7]
```diff
@@ -87,10 +87,12 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+
 public class ZKHelixManager implements HelixManager, IZkStateListener {
   private static Logger LOG = LoggerFactory.getLogger(ZKHelixManager.class);
 
   public static final String ALLOW_PARTICIPANT_AUTO_JOIN = "allowParticipantAutoJoin";
+  public static final String IS_CLOUD_ENV = "isCloudEnv";
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java]: Could we spell this out?

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+   */
+  @Override
+  public boolean validate() {
+    try {
+      Gson gson =
+          new GsonBuilder()
+              .registerTypeAdapter(AzureInstanceComputeInfo.class, new AzureComputeInfoDeserializer())
+              .create();
+      _azureInstanceComputeInfo = gson.fromJson(_response.toString(), AzureInstanceComputeInfo.class);
+    } catch (Exception e) {
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java]: To me, exception is too general. We might want to narrow it down, like JsonParseException? So we would know it is because of json parse exception, not other exception. This is just my opinion of exception handling :)

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataFetcher.java (hunk 5 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataFetcher.java]: Not only for VM. For the ones using Helix TF may have the capability using topology information for task assignment. And most of instance type of TF instances are containers.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  public AzureInstanceMetadataParser(Object response) {
+      _response = response;
+  }
+
+  /**
+   * Validate the response from Azure against the schema defined in AIMS https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service
+   * @return whether the response is a valid one from Azure
+   */
+  @Override
+  public boolean validate() {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java]: We should more validate logic there instead of just json format validate. We need to define the Azure environment required entities, FD, instance name and so on.

We need to make sure AzureInstanceComputeInfo containing all of them or at least some of them.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    return null;
+  }
+
+  class AzureComputeInfoDeserializer implements JsonDeserializer<AzureInstanceComputeInfo>
+  {
+    @Override
+    public  AzureInstanceComputeInfo deserialize(JsonElement je, Type type, JsonDeserializationContext jdc)
+        throws JsonParseException
+    {
+      JsonElement computeInfo = je.getAsJsonObject().get("compute");
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java]: Let's not hard code this here. Let's create a AzureCloudConstant class to store these kind of data.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java (hunk 8 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java]: Yeah, the exception would definitely be more specific. The current one is only a place holder. Let me put a TODO there.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceComputeInfo.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+public class AzureInstanceComputeInfo {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceComputeInfo.java]: We need an interface for CloudInstanceComputeInfo or something. This class should implement that interface.

File: helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationFetcher.java (hunk 2 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationFetcher.java]: This T type should extend CloudInstanceComputeInfo

File: helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationParser.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+   * @return whether the response is valid or not
+   */
+  boolean validate();
+
+  /**
+   * Get the value of a specific field from cloud response
+   * @param key the key of the field
+   * @return the value of the required field
+   */
+  String getValue(String key);
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationParser.java]: It should return an object such as CloudInstanceComputeInfo

File: helix-core/src/main/java/org/apache/helix/manager/zk/ParticipantManager.java [hunk 13]
```diff
@@ -97,11 +97,12 @@ public ParticipantManager(HelixManager manager, HelixZkClient zkclient, int sess
...
 
   /**
-   * Handle new session for a participang.
+   * Handle new session for a participant.
    * @throws Exception
    */
   public void handleNewSession() throws Exception {
-    joinCluster();
+
+    autoRegistration();
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/manager/zk/ParticipantManager.java]: Let's not rush for auto registration logic here. We can split it into two PRs. For this PR, let's only have the interface and class implementation. Move the registration logic to separate PR.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java (hunk 6 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java]: They can both realize our goal, however, the functions toJson() and fromJson() are more convenient in parsing the response for our use case. Let me know if multiple Json libraries will cause issue.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInformation.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+import org.apache.helix.common.cloud.CloudInformation;
+
+
+public class AzureCloudInformation implements CloudInformation {
+
+  public AzureCloudInformation() {
+
+  }
+
+  public String get(String key) {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInformation.java]: We can add more detailed function call, such as getToplogyString() later.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureSourceType.java [hunk 15]
```diff
@@ -0,0 +1,10 @@
+package org.apache.helix.cloud.azure;
+
+import org.apache.helix.common.cloud.SourceType;
+
+
+public enum AzureSourceType implements SourceType {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureSourceType.java]: Do you see any specific usage for Azure of fetching source? Otherwise, we can have a generic SourceType instead of AzureSourceType.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInformationFetcher.java [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+   */
+  @Override
+  public String get(SourceType type) {
+    AzureSourceType azureSourceType = (AzureSourceType) type;
+    switch (azureSourceType) {
+      case URL:
+        return getAzureCloudInformationFromUrl();
+      case FILE_PATH:
+        return getAzureCloudInformationFromFilePath();
+      case ZOOKEEPER_PATH:
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInformationFetcher.java]: We dont have to have these kind of entries here. We can just leave the room for extending them.

File: helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationParser.java (hunk 12 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationParser.java]: I prefer the API will be CloudInformation parseCloudInformation(List<JsonNode> data); or something else.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInformationParser.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+
+  /**
+   * Validate the response against the fields defined in Azure cloud information
+   * @return whether the response is valid
+   */
+  @Override
+  public boolean validate(CloudInformation azureCloudInformation) {
+    List<JsonNode> jsonNodes = new ArrayList<>();
+    _responses.stream().forEach(response ->  jsonNodes.add(convert(response)));
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInformationParser.java]: This is parse logic. Let's separate it out.

File: helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationParser.java [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Generic interface to parse and validate cloud information
+ */
+public interface CloudInformationParser {
+  /**
+   * Parse the cloud information from responses
+   * @return the parsed cloud information
+   */
+  CloudInformation parseCloudInformation(List<String> cloudResponses);
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationParser.java]: For interface level, I think we can make it as generic type T. For implementation, like AzureCloudInformationParser, let's make it as String.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java (hunk 6 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/cloud/azure/AzureInstanceMetadataParser.java]: I've already removed Gson.

File: helix-core/pom.xml [hunk 19]
```diff
@@ -159,6 +159,12 @@ under the License.
       <artifactId>metrics-core</artifactId>
       <version>3.2.3</version>
     </dependency>
+    <dependency>
+      <groupId>com.google.code.gson</groupId>
+      <artifactId>gson</artifactId>
+      <version>2.8.6</version>
```
[the remark this answers is not in the document]
Comment (zhangmeng916) [reply helix-core/pom.xml]: I've already removed Gson.

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java (hunk 7 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java]: Detailed implementation is moved to a separate pr, and I will address it there.

**Comment by zhangmeng916** [comment]
This PR is ready to be merged, approved by @dasahcc 

final commit message: Add Helix cloud interface and implementation skeleton methods.

File: helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationFetcher.java [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * Generic interface to fetch cloud information
+ */
+public interface CloudInformationFetcher <T extends Object> {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/common/cloud/CloudInformationFetcher.java]: Format space?

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInformationFetcher.java [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+
+  public AzureCloudInformationFetcher() {
+  }
+
+  /**
+   * Get the Azure cloud information based on source type
+   * @return Azure cloud information as a string
+   */
+  @Override
+  public String get(SourceType type) {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInformationFetcher.java]: How would we get all relevant information based solely on type? Wouldn't we need more metadata?

For example, if the type was a URL - what would be the actual url you should be reading from? What if the cloud information is spread over different types in other types of clouds? Also to support migration efforts, we might have to multiplex multiple cloud systems - will we ever support multiple types in that case (or design this in a flexible enough way for us to enable use cases like that)?

File: helix-core/src/main/java/org/apache/helix/common/cloud/SourceType.java [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+public enum SourceType {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/common/cloud/SourceType.java]: It looks like a more appropriate name for this enum would be `SourcePathType` since the enum values all refer to paths. What do you think?

Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/common/cloud/SourceType.java]: Agree there is a mismatch between name and content. Instead of adding "path", I changed the value and get rid of the "PATH". I think this is more generic.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInformationParser.java [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+
+
+public class AzureCloudInformationParser implements CloudInformationParser <String> {
+
+  public AzureCloudInformationParser() {
+  }
+
+  /**
+   * Parse Azure cloud information. It includes validating the response and populating the corresponding fields.
+   * @return whether the response is valid
```
[the remark this answers is not in the document]
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInformationParser.java]: I've updated the description. 

In previous version, we actually decomposed the parsing function and had them in the interface. However, the functions are so connected, meaning that the parsing process itself actually validates the results. Separating the logic does not make the implementation clearer or easier. 
On the other hand, in the interface design, we want to leave as much room as we can for the user to implement their logic, as for now we are considering Azure use case, we do not exactly know how other people would like to do their implementation for other cloud, e.g. gcp.  Therefore, we tend to have fewer parameters and fewer restrictions defined.

Current logic is parsing function take a list of response as input, and process them separately, together they populate fields in AzureCloudInformation. User will get the fields they need from AzureCloudInformation, and if the field is not here, an exception will be thrown.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInformationFetcher.java (hunk 21 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInformationFetcher.java]: I updated the code with constructor. The sources are input of constructor.
We do support multiple types of input in the fetcher. Based on the source type, we have different functions to retrieve the information. All the response will be collected and input to the parser. The reason we do not have 1-1 mapping from input to output is that in that way we will end up with too many types of schemas, and most of the fields are not needed. e.g. in the instance metadata response from Azure, the Json response have more than 30 fields, but we only care about 2 or 3. We will just grab the fields and populate the corresponding fields in Azure computer information file without looking at other fields.

File: helix-core/src/main/java/org/apache/helix/api/cloud/CloudInstanceInformation.java [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+  /**
+   * Get the the value of a specific cloud instance field by name
+   * @return the value of the field
+   */
+  String get(CloudInstanceField field);
+
+  enum CloudInstanceField {
+    INSTANCE_NAME,
+    FAULT_DOMAIN,
+    INSTANCE_SET_NAME
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/api/cloud/CloudInstanceInformation.java]: INSTANCES_SET_NAME

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInstanceInformation.java [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+  public AzureCloudInstanceInformation(String instanceName, String faultDomain,
+      String instanceSetName) {
+    _instanceName = instanceName;
+    _faultDomain = faultDomain;
+    _instanceSetName = instanceSetName;
+  }
+
+  @Override
+  public String get(CloudInstanceField field) {
+    switch (field) {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInstanceInformation.java]: Let's use a map to store these information. Otherwise, in the future, if we need more fields, we need to add more property.

File: helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInstanceInformation.java [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+  private String _instanceSetName;
+
+  /**
+   * Instantiate the AzureCloudInstanceInformation using each field individually.
+   * Users should use AzureCloudInstanceInformation.Builder to create information.
+   * @param instanceName
+   * @param faultDomain
+   * @param instanceSetName
+   */
+  public AzureCloudInstanceInformation(String instanceName, String faultDomain,
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInstanceInformation.java]: Same for the constructor. Let's accept a map of key-value pair as input. And make it protected, otherwise, it will be meaningless for adding builder.

File: helix-core/src/main/java/org/apache/helix/api/cloud/CloudInstanceInformation.java (hunk 24 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/api/cloud/CloudInstanceInformation.java]: The reason I keep instance is that I saw "instance group" as a name. Let's discuss this further.

File: helix-core/src/main/java/org/apache/helix/api/cloud/CloudInstanceInformationProcessor.java [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+ * under the License.
+ */
+
+import java.util.List;
+
+
+/**
+ * Generic interface to fetch and parse cloud instance information
+ */
+public interface CloudInstanceInformationProcessor<T extends Object, U extends Object> {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/api/cloud/CloudInstanceInformationProcessor.java]: Remove U type.

**Comment by zhangmeng916** [comment]
his PR is ready to be merged, approved by @dasahcc
